juju/juju
# apiserver - split the HTTP server out of the API server

## Description of change

- extracts `worker/httpserver` out of the API server. 
- introduce Authenticator and Authorizer interfaces for
  authenticating/authorizing HTTP requests
- split out a basic auth handler that wraps other
  handlers, and uses a provided Authenticator and Authorizer
- introduces a table defining endpoints in apiserver/apiserver.go,
  along with which ones require authentication, authorizers to
  use, which to track, etc.
- updates various apiserver tests to use a new base suite,
  rather than JujuConnSuite
- drops various apiserver tests which repeat testing auth
  and model UUID validation, which are orthogonal of the
  endpoints being tested.

There are numerous benefits to this (IMO), but the primary goal
is to enable rafttransport to attach to the httpserver mux, allowing
the raft object to fed into the apiserver worker.

## QA steps

Bootstrap a new controller, deploy some applications to it and relate them. Make sure upgrading the controller and hosted model works.
